File: Frontend/metrics-report.js
#!/usr/bin/env node
/**
 * Comprehensive Code Quality Metrics Report for Frontend
 * Generates detailed metrics for maintainability, complexity, and code health
 */

const { exec } = require('child_process')
const { promisify } = require('util')
const fs = require('fs').promises
const path = require('path')

const execAsync = promisify(exec)

function printSection(title) {
  console.log('\n' + '='.repeat(80))
  console.log(` ${title}`)
  console.log('='.repeat(80) + '\n')
}

async function runCommand(cmd, description) {
  try {
    console.log(`Running: ${description}...`)
    const { stdout, stderr } = await execAsync(cmd, {
      maxBuffer: 1024 * 1024 * 10, // 10MB buffer
      timeout: 300000, // 5 min timeout
    })
    return { success: true, output: stdout || stderr }
  } catch (error) {
    return { success: false, output: error.message }
  }
}

async function eslintComplexity() {
  printSection('Code Complexity (ESLint)')
  const result = await runCommand(
    'npx eslint src --format json',
    'ESLint complexity analysis'
  )

  if (result.success) {
    try {
      const data = JSON.parse(result.output)
      const complexityIssues = data
        .flatMap(file =>
          (file.messages || []).filter(msg => msg.ruleId && msg.ruleId.includes('complexity'))
        )

      console.log(`Total complexity warnings: ${complexityIssues.length}`)
      if (complexityIssues.length > 0) {
        console.log('\nTop complexity issues:')
        complexityIssues.slice(0, 10).forEach(issue => {
          console.log(`  ${issue.message} (${issue.ruleId})`)
        })
      } else {
        console.log('✓ No high complexity issues found!')
      }
    } catch (e) {
      console.log('Note: ESLint output could not be parsed as JSON')
      console.log(result.output.substring(0, 500))
    }
  }

  return result
}

async function lizardComplexity() {
  printSection('Cognitive Complexity (Lizard)')
  const result = await runCommand(
    'npx lizard src -l javascript,typescript -w -T nloc=50 -T cyclomatic_complexity=10',
    'Lizard complexity analysis'
  )

  if (result.success) {
    console.log(result.output)
  } else {
    console.log('Lizard analysis not available')
  }

  return result
}

async function codeDuplication() {
  printSection('Code Duplication Analysis (JSCPD)')
  const result = await runCommand(
    'npx jscpd src --min-lines 5 --min-tokens 50 --format "markdown"',
    'Code duplication detection'
  )

  if (result.success) {
    // Parse JSCPD output
    const output = result.output
    console.log(output.substring(0, 1000))

    if (output.includes('duplicates')) {
      console.log('\n⚠ Code duplication detected - review duplicated blocks')
    } else {
      console.log('\n✓ Low code duplication')
    }
  }

  return result
}

async function typeScriptCoverage() {
  printSection('TypeScript Type Coverage')

  // Check if tsconfig.json has necessary settings
  try {
    const tsconfigPath = path.join(__dirname, 'tsconfig.json')
    const tsconfig = JSON.parse(await fs.readFile(tsconfigPath, 'utf8'))

    console.log('TypeScript Configuration:')
    console.log(`  Strict mode: ${tsconfig.compilerOptions?.strict || false}`)
    console.log(`  No implicit any: ${tsconfig.compilerOptions?.noImplicitAny !== false}`)

    // Run type check
    const result = await runCommand('npx tsc --noEmit', 'TypeScript type check')

    if (result.success || result.output.includes('Found 0 errors')) {
      console.log('\n✓ No TypeScript errors')
    } else {
      // Count errors
      const errorMatch = result.output.match(/Found (\d+) error/)
      if (errorMatch) {
        console.log(`\n⚠ Found ${errorMatch[1]} TypeScript error(s)`)
      }
      console.log('\nFirst few errors:')
      console.log(result.output.substring(0, 500))
    }

    return result
  } catch (e) {
    console.log(`Error reading TypeScript config: ${e.message}`)
    return { success: false, error: e.message }
  }
}

async function testCoverage() {
  printSection('Test Coverage (Vitest)')

  try {
    // Check if coverage report exists
    const coveragePath = path.join(__dirname, 'coverage', 'coverage-summary.json')

    try {
      const coverageData = JSON.parse(await fs.readFile(coveragePath, 'utf8'))
      const total = coverageData.total

      console.log('Overall Coverage:')
      console.log(`  Lines: ${total.lines.pct}%`)
      console.log(`  Statements: ${total.statements.pct}%`)
      console.log(`  Functions: ${total.functions.pct}%`)
      console.log(`  Branches: ${total.branches.pct}%`)

      // Grade coverage
      const avgCoverage =
        (total.lines.pct +
          total.statements.pct +
          total.functions.pct +
          total.branches.pct) /
        4

      if (avgCoverage >= 80) {
        console.log(`\n✓ Excellent coverage (${avgCoverage.toFixed(1)}%)`)
      } else if (avgCoverage >= 60) {
        console.log(`\n⚠ Moderate coverage (${avgCoverage.toFixed(1)}%) - aim for 80%+`)
      } else {
        console.log(`\n✗ Low coverage (${avgCoverage.toFixed(1)}%) - needs improvement`)
      }

      return { success: true, coverage: avgCoverage }
    } catch (e) {
      console.log('No coverage report found. Run: npm run coverage')
      return { success: false, message: 'Coverage file not found' }
    }
  } catch (e) {
    console.log(`Error: ${e.message}`)
    return { success: false, error: e.message }
  }
}

async function linesOfCode() {
  printSection('Lines of Code Analysis')

  const result = await runCommand(
    'npx cloc src --json --exclude-dir=__tests__,node_modules',
    'LOC analysis'
  )

  if (result.success) {
    try {
      const data = JSON.parse(result.output)
      if (data.TypeScript || data.JavaScript || data.TSX) {
        console.log('Code Statistics:')
        if (data.TypeScript) {
          console.log(`  TypeScript: ${data.TypeScript.code} lines`)
        }
        if (data.TSX) {
          console.log(`  TSX (React): ${data.TSX.code} lines`)
        }
        if (data.JavaScript) {
          console.log(`  JavaScript: ${data.JavaScript.code} lines`)
        }
        console.log(`  Total: ${data.SUM.code} lines of code`)
        console.log(`  Comments: ${data.SUM.comment} lines`)
        console.log(`  Blank: ${data.SUM.blank} lines`)
      }
    } catch (e) {
      console.log('Could not parse LOC output')
    }
  } else {
    console.log(
      'Note: cloc not available. Install with: npm install -g cloc or use lizard output above'
    )
  }

  return result
}

async function generateSummary(results) {
  printSection('METRICS SUMMARY')

  console.log('Report Generated:', new Date().toLocaleString())
  console.log('\nMetrics Status:')

  Object.entries(results).forEach(([metric, result]) => {
    const status = result.success ? '✓ PASS' : '✗ FAIL'
    console.log(`  ${metric}: ${status}`)
  })

  console.log('\nRecommendations:')
  console.log('  - Complexity: Keep functions simple (CC < 10)')
  console.log('  - Duplication: Minimize code duplication (< 5%)')
  console.log('  - Test Coverage: Target 80%+ coverage')
  console.log('  - TypeScript: Use strict mode, avoid `any`')
  console.log('  - LOC: Keep components < 300 lines')

  console.log('\nFor detailed analysis:')
  console.log('  - ESLint complexity: npm run lint')
  console.log('  - Lizard: npm run metrics:complexity')
  console.log('  - JSCPD: npm run metrics:duplication')
  console.log('  - Type Coverage: npm run metrics:type-coverage')
}

async function main() {
  console.log('='.repeat(80))
  console.log(' FRONTEND CODE QUALITY METRICS REPORT')
  console.log('='.repeat(80))
  console.log(` Generated: ${new Date().toLocaleString()}`)
  console.log('='.repeat(80))

  const results = {}

  // Run all metrics
  results['ESLint Complexity'] = await eslintComplexity()
  results['Lizard Complexity'] = await lizardComplexity()
  results['Code Duplication'] = await codeDuplication()
  results['TypeScript Coverage'] = await typeScriptCoverage()
  results['Test Coverage'] = await testCoverage()
  results['Lines of Code'] = await linesOfCode()

  // Generate summary
  await generateSummary(results)

  console.log('\n' + '='.repeat(80))
  console.log(' Report complete!')
  console.log('='.repeat(80) + '\n')

  process.exit(0)
}

// Run if called directly
if (require.main === module) {
  main().catch(err => {
    console.error('Error running metrics:', err)
    process.exit(1)
  })
}

module.exports = { main }

File: Backend/scripts/analysis/metrics_report.py
#!/usr/bin/env python3
"""
Comprehensive Code Quality Metrics Report
Generates detailed metrics for maintainability, complexity, and code health
"""

import json
import subprocess
import sys
from datetime import datetime
from pathlib import Path


def run_command(cmd: list[str]) -> tuple[str, int]:
    """Run command and return output and return code"""
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, check=False, timeout=300)
        return result.stdout + result.stderr, result.returncode
    except subprocess.TimeoutExpired:
        return "Command timed out", 1
    except Exception as e:
        return f"Error: {e}", 1


def print_section(title: str):
    """Print formatted section header"""
    print(f"\n{'='*80}")
    print(f" {title}")
    print(f"{'='*80}\n")


def radon_cyclomatic_complexity():
    """Run Radon cyclomatic complexity analysis"""
    print_section("Cyclomatic Complexity (Radon)")
    output, code = run_command(["radon", "cc", ".", "-a", "-s"])
    print(output)
    return {"success": code == 0, "output": output}


def radon_maintainability_index():
    """Run Radon maintainability index analysis"""
    print_section("Maintainability Index (Radon)")
    output, code = run_command(["radon", "mi", ".", "-s"])
    print(output)
    return {"success": code == 0, "output": output}


def radon_halstead_metrics():
    """Run Radon Halstead metrics analysis"""
    print_section("Halstead Metrics (Radon)")
    output, code = run_command(["radon", "hal", ".", "-f"])
    print(output[:2000])  # Limit output
    if len(output) > 2000:
        print("\n... (output truncated) ...")
    return {"success": code == 0, "output": output[:2000]}


def radon_raw_metrics():
    """Run Radon raw metrics (LOC, LLOC, comments, etc.)"""
    print_section("Raw Metrics - Lines of Code (Radon)")
    output, code = run_command(["radon", "raw", ".", "-s"])
    print(output)
    return {"success": code == 0, "output": output}


def lizard_complexity():
    """Run Lizard complexity analysis"""
    print_section("Cognitive Complexity (Lizard)")
    output, code = run_command(["lizard", ".", "-l", "python", "-w", "-T", "nloc=50", "-T", "cyclomatic_complexity=10"])
    print(output)
    return {"success": code == 0, "output": output}


def mypy_type_coverage():
    """Run MyPy with type coverage report"""
    print_section("Type Coverage (MyPy)")
    output, code = run_command(
        [
            "mypy",
            ".",
            "--explicit-package-bases",
            "--ignore-missing-imports",
            "--no-error-summary",
            "--txt-report",
            "mypy-coverage",
            "--html-report",
            "mypy-html",
        ]
    )
    # Read type coverage statistics
    try:
        stats_file = Path("mypy-coverage/index.txt")
        if stats_file.exists():
            stats = stats_file.read_text()
            print(stats[:1000])
        else:
            print(output[:1000])
    except Exception as e:
        print(f"Could not read type coverage: {e}")
    return {"success": True, "output": "Type coverage generated"}


def test_coverage():
    """Get test coverage statistics"""
    print_section("Test Coverage (Pytest)")

    # Check if coverage.json exists
    coverage_file = Path("coverage.json")
    if coverage_file.exists():
        try:
            with open(coverage_file) as f:
                data = json.load(f)

            total_coverage = data.get("totals", {}).get("percent_covered", 0)
            print(f"Overall Test Coverage: {total_coverage:.2f}%")
            print(f"Total Statements: {data.get('totals', {}).get('num_statements', 0)}")
            print(f"Missing Statements: {data.get('totals', {}).get('num_missing', 0)}")

            # Show per-file coverage for main modules
            print("\nKey Module Coverage:")
            files = data.get("files", {})
            for filepath, file_data in list(files.items())[:10]:
                if "test" not in filepath and "venv" not in filepath:
                    coverage = file_data.get("summary", {}).get("percent_covered", 0)
                    print(f"  {filepath}: {coverage:.1f}%")

            return {"success": True, "coverage": total_coverage}
        except Exception as e:
            print(f"Error reading coverage.json: {e}")
            return {"success": False, "error": str(e)}
    else:
        print("No coverage.json found. Run: pytest --cov=. --cov-report=json")
        return {"success": False, "message": "Coverage file not found"}


def security_scan():
    """Run Bandit security scan"""
    print_section("Security Vulnerabilities (Bandit)")
    output, code = run_command(["bandit", "-r", ".", "-c", "pyproject.toml", "-f", "screen"])
    # Print summary only
    lines = output.split("\n")
    summary_start = False
    for line in lines:
        if "Run metrics:" in line or summary_start:
            summary_start = True
            print(line)
    return {"success": code == 0, "output": output}


def ruff_lint_stats():
    """Get Ruff linting statistics"""
    print_section("Linting Issues (Ruff)")
    output, code = run_command(["ruff", "check", ".", "--statistics"])
    print(output)
    return {"success": True, "output": output}


def generate_summary(results: dict):
    """Generate summary report"""
    print_section("METRICS SUMMARY")

    print("Report Generated:", datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
    print("\nMetrics Status:")

    for metric, result in results.items():
        status = "✓ PASS" if result.get("success", False) else "✗ FAIL"
        print(f"  {metric}: {status}")

    print("\nRecommendations:")
    print("  - Cyclomatic Complexity: Keep functions under 10")
    print("  - Maintainability Index: Aim for >65 (good), >85 (excellent)")
    print("  - Test Coverage: Target 80%+ for critical code")
    print("  - Type Coverage: Gradually increase to 90%+")
    print("  - Security: Address all high/medium severity issues")

    print("\nFor detailed analysis:")
    print("  - Radon: https://radon.readthedocs.io/")
    print("  - Lizard: https://github.com/terryyin/lizard")
    print("  - MyPy: https://mypy.readthedocs.io/")


def main():
    """Run all metrics and generate report"""
    print("=" * 80)
    print(" CODE QUALITY METRICS REPORT")
    print("=" * 80)
    print(f" Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print("=" * 80)

    results = {}

    # Run all metrics
    results["Cyclomatic Complexity"] = radon_cyclomatic_complexity()
    results["Maintainability Index"] = radon_maintainability_index()
    results["Halstead Metrics"] = radon_halstead_metrics()
    results["Lines of Code"] = radon_raw_metrics()
    results["Cognitive Complexity"] = lizard_complexity()
    results["Type Coverage"] = mypy_type_coverage()
    results["Test Coverage"] = test_coverage()
    results["Security Scan"] = security_scan()
    results["Linting Stats"] = ruff_lint_stats()

    # Generate summary
    generate_summary(results)

    print("\n" + "=" * 80)
    print(" Report complete!")
    print("=" * 80 + "\n")

    return 0


if __name__ == "__main__":
    sys.exit(main())
